Feature/defer to prod

beckjake opened the pull request:
resolves #2527 

### Description
Adds `--defer` and `--state` arguments, they behave as described in #2527 

I had to update hologram and make some small tweaks to support parsing the manifest from disk. The `compiled` field is now mandatory for compiled nodes, and while it may exists for a short time in a `False` state, users should never experience a situation where `compiled=False`.

The code as it is replaces all unselected nodes with entries from a previous run. I'm not sure this is as intuitive as we might want!

For example, if you have seeds you ref, and you run `dbt run --defer --state ./state`, the seeds will still point at the current run's information, because the implicit `fqn:*` selector _did_ select seeds! I think we can make some tweaks to node selection to fix this if we want, but for now I left it as-is.

Ephemeral nodes are never deferred.

### Checklist
 - [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [X] I have run this code in development and it appears to resolve the stated issue
 - [X] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

jtcohen6 commented:
This is _really_ exciting. Pulling on two of the threads you mention above:

> For example, if you have seeds you ref, and you run dbt run --defer --state ./state, the seeds will still point at the current run's information, because the implicit fqn:* selector did select seeds! I think we can make some tweaks to node selection to fix this if we want, but for now I left it as-is.

This is the expected behavior IMO, since it ensures that `dbt run --defer --state ./state` == `dbt run`. If someone wanted to defer seed files _only_ (which would surprise me), I'd expect them to do something a little more explicit like `dbt run --defer --state ./state --exclude my_one_seed my_other_seed`.

> Ephemeral nodes are never deferred.

To clarify my understanding: Ephemerals are passthroughs, and so any of _their_ references to non-ephemeral, unselected models _will_ be deferred. To that end, it doesn't matter whether the ephemeral models themselves are included or excluded from the set of selected nodes.

beckjake commented:
Yeah. I played around with a bit and deferring to ephemeral models really violates my expectations about `defer`: in particular that it only defers to the file for "how do I refer to this model's existence in the database?". That's a nonsensical question for ephemeral models: it's their contents that are important downstream, not their names!

beckjake commented:
I've rebased onto the newly-merged `dev/marian-anderson` and added some more tests. I'll mark this as ready for review as soon as tests are passing, but some local spot testing looked ok

jtcohen6 commented on core/dbt/contracts/graph/manifest.py:
```diff
@@ -1011,25 +1011,52 @@ def resolve_doc(
+            ):
+                merged.add(unique_id)
+                self.nodes[unique_id] = node
+
```
Where is the logic to flip on the `deferred` flag? I see that `deferred: True` in several of the test cases. When I ran with deferral locally and check the manifest afterward, `deferred` is always false.

jtcohen6 commented on core/dbt/task/run.py:
```diff
@@ -245,6 +251,34 @@ def __init__(self, args, config):
+    def _get_state_path(self) -> Path:
+        if self.args.state is not None:
+            return self.args.state
+        elif ARTIFACT_STATE_PATH is not None:
```
Nice! Glad to see that `ARTIFACT_STATE_PATH` and `--state` are mutually substitutable, to support Cloud deployment.

I _think_ the conclusion of [this conversation](https://github.com/fishtown-analytics/dbt-cloud/issues/1497#issuecomment-663203559) was to add a separate environment variable for deferral as well. Let me know if you disagree.

beckjake replied:
I did that! It's set up differently because it's a boolean rather than a value, but it's defined in `dbt/flags.py`. It being set/unset informs the default value for the `--defer`/`--no-defer` argument to change.
